### PR TITLE
Improvement: 3093 Dashboard Services panel

### LINKED ISSUE #3093: MetalK8s UI Overview page Services Panel
**Component**: UI, monitoring

**Why this is needed**:
We want to display the health of the grouped services.

**What should be done**:

image

~~We would require the final design from @Cuervino @thomasdanan about the interaction, more specifically, when hovering on the status what should we display on the tooltip.~~
~~-  We currently blocked by the group alert feature. (so this ticket is only about UI layout)~~
(Edited: Not anymore blocked)
~~- This component should include a layout property and stay independent to be reused by other applications at runtime.~~
(Edited: We don't require this anymore since we will import the alerthooks from shell UI through the federated hook)

Interaction design:
https://www.figma.com/proto/N2jOAK44dTSiBclDjXzz6a/Next_Gen_Design?node-id=301%3A116&viewport=1035%2C-840%2C0.23489561676979065&scaling=min-zoom&page-id=180%3A3495

**Implementation proposal** (strongly recommended):
We need to restyle the tooltip in core-ui to meet the interaction design. 

**Test plan**:
TBD

### PR DESCRIPTION
**Pull request by alexis-ld**
**Component**: dashboard

**Context**:  #3093 

**Summary**: Adds a Services panel to the dashboard

https://user-images.githubusercontent.com/3480526/127501863-1663e0bd-936c-4812-8854-4f1b2bc2557b.mov

Closes: #3093

### CODE REVIEW

File: ui/src/components/DashboardServices.js [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        break;
+      case STATUS_CRITICAL:
+        color = theme.statusCritical;
+        break;
+      default:
+        color = theme.statusHealthy;
+    }
+    return color;
+  }};
+`;
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: could be removed

File: ui/src/components/DashboardServices.js [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+            height: '100%',
+          }}
+          overlay={
+            <NonHealthyPopUp>
+              <NonHealthyPopUpTitle>View details</NonHealthyPopUpTitle>
+              <NonHealthyPopUpItem>
+                <label>Severity</label>
+                <NonHealthyPopUpSeverity status={status}>
+                  {status}
+                </NonHealthyPopUpSeverity>
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: Could benefit from the New `Text` component  in core-ui.  You may want to check the `statusText` in storybook

File: ui/src/components/DashboardServices.js [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+          <CircleStatus status={status} />
+          <ServiceItemLabel>{label}</ServiceItemLabel>
+        </ServiceItemLabelWrapper>
+      </ServiceItemElement>
+    );
+  else
+    return (
+      <NonHealthyServiceItemElement>
+        <Tooltip
+          // Right placement to avoid Z index issues between left sidebar and tooltip or out of screen tooltip displays
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js:123]: can be removed since we already fixed the issue ;)

File: ui/src/components/DashboardServices.js [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+  self-align: flex-end;
+`;
+
+const ServiceItem = ({
+  label,
+  status,
+  alerts,
+}: {
+  label: string,
+  status: string,
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: I believe the type of status can be more specific. 

We have already typed the Status in `AlertProvider`
> export type Status = 'healthy' | 'warning' | 'critical';

File: ui/src/components/DashboardServices.js [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+  alerts: Alert[],
+}) => {
+  let date = null;
+
+  const getStartDate = (alerts: Alert[]) => {
+    // Check we only have alerts and set startsAt properties
+    const filtered = alerts.filter((item) => item && item.startsAt);
+    // Sort to make sure we have oldest alert first
+    filtered.sort((a, b) => new Date(a.startsAt) - new Date(b.startsAt));
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: it's good thinking, however, I don't think it's necessary ;)
In theory, at the current time, the same alert should not be triggered more than once.

File: ui/src/components/DashboardServices.js [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+      </NonHealthyServiceItemElement>
+    );
+};
+
+const DashboardServices = () => {
+  const intl = useIntl();
+  const alertsLibrary = useAlertLibrary();
+
+  // K8s Master
+  const k8sAlerts = useHighestSeverityAlerts(
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: In this specific case, we only gather two alerts: "K8sMasterServicesAtRisk" and "K8sMasterServicesDegraded"
and then we filter out the lower severity, so in the end, we should have at most one alert or an empty array. 
```suggestion
  const k8sHighestSeverityAlert = useHighestSeverityAlerts(
```

File: ui/src/components/DashboardServices.js [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+`;
+
+const ServiceItem = ({
+  label,
+  status,
+  alerts,
+}: {
+  label: string,
+  status: string,
+  alerts: Alert[],
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js:106]: IMO, we should have at most one alert, please see the explanation below :)
suggestion:
`alert: Alert,`

File: ui/src/components/DashboardServices.js [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+          // Right placement to avoid Z index issues between left sidebar and tooltip or out of screen tooltip displays
+          placement="top"
+          overlayStyle={{
+            // em sizing to handle font-size change on large screens
+            width: '20em',
+            height: '100%',
+          }}
+          overlay={
+            <NonHealthyPopUp>
+              <NonHealthyPopUpTitle>View details</NonHealthyPopUpTitle>
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: I think we should translate this as well.

File: ui/src/components/DashboardServices.js [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+              <NonHealthyPopUpItem>
+                <label>Severity</label>
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: missing translation.

File: ui/src/components/DashboardServices.js [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+                <NonHealthyPopUpSeverity status={status}>
+                  {status}
+                </NonHealthyPopUpSeverity>
+              </NonHealthyPopUpItem>
+              {
+                // Checking date validity
+                date instanceof Date && !isNaN(date) && (
+                  <NonHealthyPopUpItem>
+                    <label>Start</label>
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: ideally, we should translate start as well.

File: ui/src/components/DashboardServices.js [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+                    <div>
+                      {`${date.toISOString().split('T')[0]}
+              ${date.toISOString().split('T')[1].slice(0, 8)}`}
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: I think second is not necessary.
so you could benefit from`formatDateToMid1` in utils.js

File: ui/src/components/__TEST__/util.js [hunk 12]
```diff
@@ -58,15 +62,17 @@ const AllTheProviders = ({ children }) => {
...
-    <IntlProvider locale="en" messages={translations_en}>
-      <Provider store={store}>
-        <QueryClientProvider client={queryClient}>
-          <AlertProvider>
-            <ThemeProvider theme={theme}>{children}</ThemeProvider>
-          </AlertProvider>
-        </QueryClientProvider>
-      </Provider>
-    </IntlProvider>
+    <Router history={history}>
```
Comment (ChengYanJin) [inline ui/src/components/__TEST__/util.js:65]: can you please explain to me why we need this Wrapper?

File: ui/src/components/DashboardServices.js [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+} from '@scality/core-ui/dist/style/theme';
+
+import type { Alert } from '../services/alertUtils';
+import { PageSubtitle } from '../components/style/CommonLayoutStyle';
+import {
+  useAlertLibrary,
+  useHighestSeverityAlerts,
+  highestAlertToStatus,
+} from '../containers/AlertProvider';
+import CircleStatus from './CircleStatus';
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js:21]: I thought we already move it core-ui, but apparently not.. ;) 
Not the scope for this PR will do it in another PR.

**Review by ChengYanJin — CHANGES_REQUESTED** [review]

File: ui/src/components/__TEST__/util.js (hunk 12 above)
Comment (alexis-ld) [reply ui/src/components/__TEST__/util.js:65]: @ChengYanJin , because we use react-router-dom's `Link` in the component, without the Router wrapper this makes the test suite fail because react-router component cannot be rendered outside a Router context.

Here we create an in-memory fake history to be provided to the test suite.

File: ui/src/components/DashboardServices.js [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+              <NonHealthyPopUpTitle>
+                {intl.formatMessage({ id: 'view_details' })}
+              </NonHealthyPopUpTitle>
+              <NonHealthyPopUpItem>
+                <label>{intl.formatMessage({ id: 'severity' })}</label>
+                <StatusText status={status}>{status}</StatusText>
+              </NonHealthyPopUpItem>
+              {
+                // Checking date validity
+                date && (
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]:
```suggestion
                alert[0].startsAt && (
```
if we want to make sure we have a startsAt property, we can check here.

File: ui/src/components/DashboardServices.js [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+  const alertsLibrary = useAlertLibrary();
+
+  // K8s Master
+  const k8sHighestSeverityAlert = useHighestSeverityAlerts(
+    alertsLibrary.getK8SMasterAlertSelectors(),
+  );
+  const k8sStatus = highestAlertToStatus(k8sHighestSeverityAlert);
+
+  // Bootstrap
+  const bootstrapAlerts = useHighestSeverityAlerts(
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]:
```suggestion
  const bootstrapHighestSeverityAlert = useHighestSeverityAlerts(
```
same for the rest...

File: ui/src/components/DashboardServices.js [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+  label: string,
+  status: Status,
+  alerts: Alert[],
+}) => {
+  let date = null;
+  const intl = useIntl();
+
+  const getStartDate = (alerts: Alert[]) => {
+    // Check we only have alerts and set startsAt properties
+    const filtered = alerts.filter((item) => item && item.startsAt);
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js]: I think the filter is not necessary here, since what you want is only a check to make sure we do have the startsAt property.

File: ui/src/components/DashboardServices.js [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+  if (!alerts.length && status === STATUS_HEALTH)
+    return (
+      <ServiceItemElement aria-label={label}>
+        <ServiceItemLabelWrapper>
+          <CircleStatus status={status} />
+          <ServiceItemLabel>{label}</ServiceItemLabel>
+        </ServiceItemLabelWrapper>
+      </ServiceItemElement>
+    );
+  else
```
Comment (ChengYanJin) [inline ui/src/components/DashboardServices.js:119]: if you want to be super strict,  but I think it's not necessary.. 
```suggestion
  else if(alerts.length === 1  && status !== STATUS_HEALTH)
```
and in the end, if we are not in any of the two scenarios, 
else{
 console.error(` The Service ${label} status has encountered an expected scenario.`)
}

**Review by ChengYanJin — APPROVED** [review]
LGTM!
